---
title: "Primer_pair_wrapper"
author: "Ramon Gallego"
date: "11/19/2021"
output: html_document
params:
  metadata: 
    value: ../data/primers.csv
  Num.alignments:
    value: 5000
  outputfolder:
    value: ../test
  NCBI_api:
    value: !r Sys.getenv("NCBI_API_KEY")
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Primer pair wrapper

Given a set of primers, and a target of alignments, return the sequences that match, and the taxonomical list of those sequences

```{r libraries, echo = F}
library(tidyverse)
library (primerTree)
```

### Capture the parameters



```{r }
dir.create(params$outputfolder)
metadata <- read_csv(params$metadata)

## Check metadata has the columns we need

reqs <-  c("Locus","Sequence_PrimerF","Sequence_PrimerR","Name_PrimerF","Name_PrimerR" )

if (sum(reqs %in% colnames(metadata)) < length(reqs)  ){
    message <- cat(paste(reqs, collapse = '\n'))
    knitr::knit_exit(append = paste0(" ERROR: Initial metadata is missing some of the key column names:
                                     ",
                                     paste(reqs, collapse = '\n'),
                                     "
                                     Change the column names / add the infromation to your csv file",
                                     sep = '\n' ))
}

# Check there are unique values for Locus

metadata %>%
  group_by(Locus) %>% 
  tally() %>% 
  pull(n) -> counts
if (max(counts)>1 ){
    message <- cat(paste(reqs, collapse = '\n'))
    knitr::knit_exit(append = " ERROR: Initial metadata has repeated values for `Locus`
                                     Make sure file is formatted correctly")
}

```

```{r function, echo=F }
Recover_taxonomy_list <- function(Primer.pair) {
  
  Primer.pair$taxonomy %>%
    left_join(Primer.pair$BLAST_result,by="gi")  %>%
    distinct(order, genus, species, accession,taxId, mismatch_forward, mismatch_reverse) %>%
    arrange(desc(genus))
}
```

```{r do the thing}

metadata %>% 
  group_by(Locus) %>% 
  nest() %>% 
  mutate(primer.output = map2 (data,Locus,  ~search_primer_pair(forward = .x$Sequence_PrimerF,
                                                        reverse = .x$Sequence_PrimerR,
                                                        name = .y,
                                                        num_aligns = params$Num.alignments,
                                                        num_permutations = 25,
                                                        simplify = TRUE,
                                                        clustal_options = list(exec = "clustalo",
                                                                               quiet = TRUE, original.ordering = TRUE),
                                                        distance_options = list(model = "N", pairwise.deletion = T),
                                                        api_key = params$NCBI_api,
                                                        .parallel = FALSE,
                                                        .progress = "none"))
) -> primer.output

write_rds(primer.output, file = file.path(params$outputfolder, "primer_search.rds"))

```

Get the taxonomy

```{r}
primer.output %>% 
  mutate(taxonomy.list = map(primer.output, Recover_taxonomy_list)) -> Cet12S.NCBI.PCR.list

Cet12S.NCBI.PCR.list %>%
  select(Locus, data, taxonomy.list) %>% 
  unnest(data) %>% 
  unnest(taxonomy.list) %>% 
  write_csv(file.path(params$outputfolder, "taxonomy.recovered.csv"))
```

